File: src/main/java/org/cubewhy/celestial/files/ProxyConfig.java
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.files;

import cn.hutool.core.net.url.UrlBuilder;
import cn.hutool.core.net.url.UrlQuery;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.net.*;
import java.nio.charset.StandardCharsets;

public class ProxyConfig extends ConfigFile {

    @Getter
    public static class Mirror {

        private final String host;
        private final int port;

        public Mirror(@NotNull String address) {
            host = address.split(":")[0];
            port = Integer.parseInt(address.split(":")[1]);
        }
    }

    public ProxyConfig(File file) {
        super(file);
        this.initValue("state", new JsonPrimitive(false));
        this.initValue("proxy", "http://127.0.0.1:8080");
        this.initValue("mirror", new JsonObject());
    }

    public ProxyConfig useProxy(@NotNull URL address) {
        this.setValue("proxy", address.toString());
        this.save();
        return this;
    }
    
    public ProxyConfig setState(boolean state) {
        this.setValue("state", state);
        return this;
    }
    
    public boolean getState() {
        return this.getValue("state").getAsBoolean();
    }

    public Proxy getProxy() throws MalformedURLException {
        URL address = new URL(this.getProxyAddress());
        return new Proxy(getType(address.getProtocol()), new InetSocketAddress(address.getHost(), address.getPort()));
    }

    public String getProxyAddress() {
        return this.getValue("proxy").getAsString();
    }

    private Proxy.Type getType(@NotNull String protocol) {
        return switch (protocol) {
            case "http" -> Proxy.Type.HTTP;
            case "socks" -> Proxy.Type.SOCKS;
            default -> throw new IllegalStateException("Unexpected value: " + protocol);
        };
    }

    public URL useMirror(@NotNull URL src) throws MalformedURLException, URISyntaxException {
        String host = src.getHost();
        int port = src.getPort();
        if (port == -1) {
            port = src.getDefaultPort();
        }
        String completed = host + ":" + port;
        if (this.getConfig().getAsJsonObject("mirror").has(completed)) {
            Mirror mirror = getMirror(completed);
            return new URL(src.getProtocol(), mirror.host, mirror.port, src.getFile());
        }
        return src;
    }

    public Mirror getMirror(String address) {
        JsonObject json = getValue("mirror").getAsJsonObject();
        return new Mirror(json.get(address).getAsString());
    }
}

File: src/main/kotlin/org/cubewhy/celestial/files/ProxyConfig.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */
package org.cubewhy.celestial.files

import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.URL

class ProxyConfig(file: File?) : ConfigFile(file!!) {
    class Mirror(address: String) {
        val host: String = address.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        val port: Int = address.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].toInt()
    }

    init {
        this.initValue("state", JsonPrimitive(false))
        this.initValue("proxy", "http://127.0.0.1:8080")
        this.initValue("mirror", JsonObject())
    }

    fun useProxy(address: URL): ProxyConfig {
        this.setValue("proxy", address.toString())
        this.save()
        return this
    }

    fun setState(state: Boolean): ProxyConfig {
        this.setValue("state", state)
        return this
    }

    val state: Boolean
        get() = this.getValue("state").asBoolean

    val proxy: Proxy?
        get() {
            val address = URL(this.proxyAddress)
            return if (state) Proxy(getType(address.protocol), InetSocketAddress(address.host, address.port)) else null
        }

    private val proxyAddress: String
        get() = this.getValue("proxy").asString

    private fun getType(protocol: String): Proxy.Type {
        return when (protocol) {
            "http" -> Proxy.Type.HTTP
            "socks" -> Proxy.Type.SOCKS
            else -> throw IllegalStateException("Unexpected value: $protocol")
        }
    }

    fun useMirror(src: URL): URL {
        val host = src.host
        var port = src.port
        if (port == -1) {
            port = src.defaultPort
        }
        val completed = "$host:$port"
        if (config.getAsJsonObject("mirror").has(completed)) {
            val mirror = getMirror(completed)
            return URL(src.protocol, mirror.host, mirror.port, src.file)
        }
        return src
    }

    private fun getMirror(address: String?): Mirror {
        val json = getValue("mirror").asJsonObject
        return Mirror(json[address].asString)
    }

    fun applyMirrors(mirrors: JsonObject?): ProxyConfig {
        log.info("apply mirrors")
        this.setValue("mirror", mirrors)
        return this
    }

    fun addMirror(source: String, mirror: String) {
        this.config["mirror"].asJsonObject.addProperty(source, mirror)
    }

    fun hasMirror(source: String): Boolean {
        return this.config["mirror"].asJsonObject.has(source)
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(ProxyConfig::class.java)
    }
}
